warehouse/query-core/src/main/java/datawave/next/CountQueryIterator.java

```
@@ -0,0 +1,170 @@
+
+import datawave.core.iterators.ResultCountingIterator;
+import datawave.query.iterator.QueryIterator;
+import datawave.query.iterator.QueryOptions;
+import datawave.query.jexl.JexlASTHelper;
+
+/**
+ * An alternate {@link QueryIterator} that provides counts for simple queries.
+ * <p>
+ * This was an attempt an at optimized count query for the field index.
+ */
+public class CountQueryIterator implements SortedKeyValueIterator<Key,Value> {
+
+    private final Logger log = LoggerFactory.getLogger(CountQueryIterator.class);
+
+    private Range range;
+    private ASTJexlScript script;
+    private Set<String> datatypeFilter;
+    private LongRange timeFilter;
+
+    private Set<String> indexedFields;
+
+    private SortedKeyValueIterator<Key,Value> source;
+    private Map<String,String> options;
+    private IteratorEnvironment env;
+
+    private Key tk;
+    private Value tv = new Value();
+
+    public CountQueryIterator() {}
+
+    public CountQueryIterator(CountQueryIterator other, IteratorEnvironment env) {
+        this.source = other.source.deepCopy(env);
+        this.options = other.options;
+        this.env = other.env;
+
+        this.range = other.range;
+        this.script = other.script;
+        this.datatypeFilter = other.datatypeFilter;
+        this.timeFilter = other.timeFilter;
+    }
+
+    @Override
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+        this.source = source;
+        this.options = options;
+        this.env = env;
+
+        // need the query, datatype filter, start and end dates (for time filter)
+
+        if (options.containsKey(QueryOptions.QUERY)) {
+            String opt = options.get(QueryOptions.QUERY);
+
+            try {
+                this.script = JexlASTHelper.parseAndFlattenJexlQuery(opt);
+            } catch (ParseException e) {
+                throw new RuntimeException(e);
+            }
+        }
```

If no QueryOptions.QUERY is passed this should throw an error

---

correct! done in the next commit, thanks